### PR TITLE
Llama2 7B Quantization Examples

### PR DESCRIPTION
**Pull request by Satrat**
Creating a new examples folder, with initial examples for llama7b using ultrachat200k
* W4A16 channelwise (notebook and python script)
* W8A8 channelwise weights, dynamic per token activations (python script)
* 2:4 sparsity -> finetuning -> W4A16 channelwise quantization (README and python script) 

### Models

Storing the model outputs on network under `/network/sadkins`
* llama7b_w4a16_channel_compressed
* llama7b_w8a8_channel_dynamic_compressed

### Eval Results
```
sparseml.evaluate /network/sadkins/llama1.1b_W4A16_channel_compressed -d wikitext -i lm-evaluation-harness
```

* `zoo:llama2-7b-ultrachat200k_llama2_pretrain-base` (baseline): 10.10
* `llama7b_w4a16_channel_compressed`: 11.09
* `llama7b_w8a8_channel_dynamic_compressed`: 10.17

## Missing
* 2:4 quantized run is waiting on the GPTQ UX changes to merge, currently the originally sparsity is not respected by GPTQ. So far I've just confirmed the script completes with a 1.1b model.
* Waiting on group quantization correctness fixes before validating a 7b grouped example

### CODE REVIEW

**Review by rahul-tuli — APPROVED** [review]

File: examples/llama7b_sparse_quantized/README.md
Comment (dbogunowicz) [inline examples/llama7b_sparse_quantized/README.md:5]:
```suggestion
the ultrachat200k dataset.
```

File: examples/llama7b_sparse_quantized/README.md
Comment (dbogunowicz) [inline examples/llama7b_sparse_quantized/README.md:37]: Is that right? So the `compress` pathway does not use `compressed-tensors` when saving the optimized model to `output_dir = "output_llama7b_2:4_w4a16_channel_compressed"`?

File: src/sparseml/transformers/compression/quantization_format.py
Comment (dbogunowicz) [inline src/sparseml/transformers/compression/quantization_format.py:60]: don't we have this func in `compressed-tensors` already?

File: examples/llama7b_w4a16_quantization.ipynb
Comment (dbogunowicz) [inline examples/llama7b_w4a16_quantization.ipynb:1]: <p>I think it would be nice to have a chance to validate those notebooks locally, like I do in <code>compressed-tensors</code> . This would mean setting up a flow that lets you run notebooks and check whether they terminate without an error.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/neuralmagic/sparseml/pull/2285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (dbogunowicz) [inline examples/llama7b_w4a16_quantization.ipynb:1]: <p>Maybe important to mention that vLLMQuantizationModifier is in general a PTQ method (for those who are quant-savvy)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/neuralmagic/sparseml/pull/2285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

File: examples/llama7b_sparse_quantized/README.md
Comment (Satrat) [reply examples/llama7b_sparse_quantized/README.md:37]: So we only save in compressed format if `save_compressed=True` is passed to sparseml.train/oneshot/apply/compress. This is really a naming issue we should address: we decided to call the multi_stage pathway "compress()" before compressed-tensors existed. Its a completely different pathway from ModelCompressor.compress() which actually compresses the model on disk. sparseml.compress() just runs oneshot and or finetuning. I'll update it to use the apply() function instead which is equivalent and change the output_dir name to make this less confusing.

I chose for this example to save in an uncompressed format to avoid running compression on all the checkpoints

File: src/sparseml/transformers/compression/quantization_format.py
Comment (Satrat) [reply src/sparseml/transformers/compression/quantization_format.py:60]: Link?? I didn't think we had anything similar

File: examples/llama7b_w4a16_quantization.ipynb
Comment (Satrat) [reply examples/llama7b_w4a16_quantization.ipynb:1]: Yeah thats a good idea, though we probably don't want to run these 7b models as part of our regular testing. Maybe we could set them up to run weekly?

Comment (bfineran) [inline examples/llama7b_w4a16_quantization.ipynb:1]: <p>definitely should update to simplify after the UX fixes - ideally we don't need to expose too much about recipes/modifiers here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/neuralmagic/sparseml/pull/2285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (bfineran) [inline examples/llama7b_w4a16_quantization.ipynb:1]: <p>nit: name test_stage</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/neuralmagic/sparseml/pull/2285/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

**Review by bfineran — APPROVED** [review]
lgtm pending comments

File: examples/llama7b_w4a16_quantization.ipynb
Comment (Satrat) [reply examples/llama7b_w4a16_quantization.ipynb:1]: Yup, once the UX PRs get merged in I'll update the examples

**Review by bfineran — APPROVED** [review]

**Pull request merged**
